File: java/Hot100/Array/FirstMissingPositive.java
package Hot100.Array;

/**
 * Given an unsorted integer array nums. Return the smallest positive integer that is not present in nums.

 * You must implement an algorithm that runs in O(n) time and uses O(1) auxiliary space.
 */
public class FirstMissingPositive {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            // put the positive numbers into the right places
            // keep swaping till i has the right value
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                swap(nums, nums[i] - 1, i);
            }
        }

        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }

        return n + 1;
    }

    private void swap(int[] nums, int index1, int index2) {
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }
}

File: python/array/first_missing_positive.py
# Given an unsorted integer array nums. Return the smallest positive integer that is not present in nums.
# You must implement an algorithm that runs in O(n) time and uses O(1) auxiliary space.
class Solution:
    def firstMissingPositive(self, nums: list[int]) -> int:
        n = len(nums)

        # Place each value x in its correct index x-1 if 1 <= x <= n.
        for i in range(n):
            while 1 <= nums[i] <= n and nums[nums[i] - 1] != nums[i]:
                target = nums[i] - 1
                nums[i], nums[target] = nums[target], nums[i]

        # The first index i where nums[i] != i + 1 is the answer.
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1

        return n + 1
